Loading - Add component

veteran29 opened the pull request:
Add new loading system with API that allows other components to register their own handlers which tell us if the component is ready.

Removes bunch of calls to legacy MF stuff.

veteran29 committed:
Remove legacy info panel and welcome screen logic

veteran29 committed:
Add VGM_DisplayLoading

veteran29 committed:
Add loading system

Spoffy commented on game/functions/core/client/loading/fn_loading_postInit.sqf:
```diff
@@ -0,0 +1,48 @@
+
+    Example(s):
+        [parameter] call vgm_X_fnc_component_myFunction
+ */
```
Docs 🙂

Spoffy commented on game/functions/core/client/loading/fn_loading_postInit.sqf:
```diff
@@ -0,0 +1,48 @@
+        uiSleep 0.1;
+    };
+
+    "vgm_loading" call BIS_fnc_endLoadingScreen;
```
Having a bit of difficulty understanding the behaviour here. I read this as:
* Call all handlers to start loading, delete if it's done.

* Update text with something like:
 ```
 Loading.... System1
 System2
 System3
 ```
 
 * Update loading bar with a % based on how many handlers haven't yet returned true.
 
 
Does that sound right?

Spoffy replied:
It's worth adding some kind of fallback to this function. If a script errors, we'll get locked in the loading screen and have to restart Arma I believe.

Spoffy replied:
Can you shoot while on the loading screen? I vaguely recall that being the reason we originally disabled player simulation while loading.

veteran29 replied:
> Does that sound right?

yes

> Can you shoot while on the loading screen? I vaguely recall that being the reason we originally disabled player simulation while loading.

No, if you `BIS_fnc_startLoadingScreen` during `initFunctions` `postInit` it will be added to the "stack" of loading screens. Control over player avatar is given after all of them are done. I think you needed that as you did not use BI func and fired the loading screen at some later point of execution.

veteran29 replied:
> It's worth adding some kind of fallback to this function.

I think we need a timeout, but what should it do? In "production" it should probably kick you from the game telling you that something went wrong. Not sure about dev, maybe kick if no debug console and print an error and close loading screen if player has access to it?

veteran29 replied:
Added kick for standard user and error message for admins/debug console users after a 30s timeout.

veteran29 committed:
Docs

veteran29 committed:
Add timeout to loading screen

Admin will be let into the mission with error popup, standard user will be kicked.

veteran29 committed:
Prevent handlers from breaking the script by spawning them

veteran29 committed:
Animate "Loading..." dots, increase timeout time

veteran29 committed:
Merge remote-tracking branch 'origin/development' into loading
